Tribler/community/dht/community.py

```diff
@@ -0,0 +1,373 @@
+        self.decode_map.update({
+            chr(7): self.on_ping_request,
+            chr(8): self.on_ping_response,
+            chr(9): self.on_store_request,
+            chr(10): self.on_store_response,
+            chr(11): self.on_find_request,
+            chr(12): self.on_find_response,
+        })
+
+        self.logger.info('DHT community initialized (peer mid %s)', self.my_peer.mid.encode('HEX'))
+
+    @inlineCallbacks
+    def unload(self):
+        self.request_cache.clear()
+        yield super(DHTCommunity, self).unload()
+
+    @property
+    def my_node_id(self):
+        return calc_node_id(self.my_peer.address[0], self.my_peer.mid)
+
+    def send_message(self, address, message_id, payload_cls, payload_args):
+        global_time = self.claim_global_time()
+        auth = BinMemberAuthenticationPayload(self.my_peer.public_key.key_to_bin()).to_pack_list()
+        payload = payload_cls(*payload_args).to_pack_list()
+        dist = GlobalTimeDistributionPayload(global_time).to_pack_list()
+        packet = self._ez_pack(self._prefix, message_id, [auth, dist, payload])
+        return self.endpoint.send(address, packet)
+
+    def on_introduction_request(self, source_address, data):
+        super(DHTCommunity, self).on_introduction_request(source_address, data)
+        auth, _, _ = self._ez_unpack_auth(IntroductionRequestPayload, data)
+        # Filter out trackers
+        if source_address not in self.network.blacklist:
+            node = Node(auth.public_key_bin, source_address)
+            self.on_node_discovered(node)
+
+    def on_introduction_response(self, source_address, data):
+        super(DHTCommunity, self).on_introduction_response(source_address, data)
+        auth, _, _ = self._ez_unpack_auth(IntroductionResponsePayload, data)
+        # Filter out trackers
+        if source_address not in self.network.blacklist:
+            node = Node(auth.public_key_bin, source_address)
+            self.on_node_discovered(node)
+
+    @inlineCallbacks
+    def on_node_discovered(self, node):
+        if not self.routing_table.has(node):
+            node = self.routing_table.add(node)
+
+            if node:
+                self.logger.info('Added node %s to the routing table', node.address)
+
+                # Ping the node in order to determine RTT
+                yield self.ping(node)
+
+                # Check if we need to move data to the new node
+                for key, values in self.storage.data.iteritems():
+                    if distance(key, self.my_node_id) > distance(key, node.id):
+                        self.store_on_nodes(key, [v for _, _, v in values], [node])
+
+    def ping_all(self):
+        deferreds = []
+        now = time.time()
+        for bucket in self.routing_table.trie.values():
+            for node in bucket.nodes.values():
+                if node.last_response + PING_INTERVAL <= now:
+                    deferreds.append(self.ping(node).addErrback(lambda _: None))
```

Comment:
It is correct that you do not want to do anything when a ping to a node fails?

Reply:
It is. When a ping fails, Node.failed will be incremented, which will eventually lead to the Node being removed from the routing table.